Add regression test for editor JS crash caused by rtlcss parsing exception

fullofcaffeine opened the pull request:
## Description

Issue: https://github.com/WordPress/gutenberg/issues/29250.

Add regression test to prevent issues related to RTL being active at the WP-admin level.*

_*it seems you can activate RTL in two places: wp-admin and in the Gutenberg editor. The error we're trying to avoid a regression for only happens when RTL is active at the "WP admin level" (someone please correct me if this is the wrong mental model / I'm using the wrong term here)._

After investigating where `rtlcss` is parsing the CSS (see: https://github.com/ItsJonQ/g2/blob/main/packages/create-styles/src/create-compiler/plugins/rtl.js#L38), the challenge was then making sure that `document.dir = 'rtl'` ran before that. Played with several approaches, including using Puppeteer's `evaluateOnNewDocument` to execute the snippet, but none of them worked. 

Had to do some gymnastics to make sure the snippet ran before the corresponding js. With that test, I am able to reproduce the error in GB <= 10.0.0. Still need to clean it up, and it's less than perfect _(it effectively reloads the script after the page already loaded, while making sure the dir RTL property is set in the document)_ but it seems this approach is feasible. 

## How to test

1. Easiest way is to checkout `v10.0.0` (or revert commit `2f93c432a3`), then `npm ci && npm run build` and start a wp-env instance from there. Finally, in another Gutenberg copy, check out this branch (`add/regression-test-rtlcss-parser-comment-crash`) and run this test _(you might also call `only` in the `it` to restrict the run to the single test added in this PR, if you'd like)_. The `loads editor without errors when RTL is set at the WP admin level` test should fail 🔴:

![expect](https://user-images.githubusercontent.com/81248/109231946-01d89f80-778d-11eb-8064-68bd432c830e.png)


2. Now, where you had `v10.0.0` checkout, checkout `v10.0.2` (or `master`) instead, run `npm ci && npm run build` and go back to the other Gutenberg instance, and run the test again. It should pass 🟢:

![pass](https://user-images.githubusercontent.com/81248/109232624-33059f80-778e-11eb-8bfa-6efc4f828467.png)

## Types of changes

New E2E test.

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

fullofcaffeine committed:
Add regression test for editor JS crash caused by rtlcss parsing exception

ockham commented:
Wondering if we could a different puppeteer helper here, such as [`page.evaluateOnNewDocument`](https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pageevaluateonnewdocumentpagefunction-args), to avoid some of the clunkiness of the script removal/appending :grimacing:

fullofcaffeine commented:
> Wondering if we could a different puppeteer helper here, such as page.evaluateOnNewDocument, to avoid some of the clunkiness of the script removal/appending grimacing

This was the first approach I tried. I've set the `dir` attribute on `document` from this callback but it ended up not being available when the actual document loaded. It seems the objects are different. I'll double-check nevertheless.

fullofcaffeine commented:
> Wondering if we could a different puppeteer helper here, such as [`page.evaluateOnNewDocument`](https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pageevaluateonnewdocumentpagefunction-args), to avoid some of the clunkiness of the script removal/appending grimacing

I tried again and setting the attribute on `evaluateOnNewDocument` before calling `createNewPost()`, but it doesn't seem to do the trick, the `dir` attribute doesn't persist over to the actual page load for `post-new.php`:


https://user-images.githubusercontent.com/81248/109205074-b8c42380-776b-11eb-81c1-e1f1ff6bc96e.mp4

![Screenshot from 2021-02-25 13-02-44](https://user-images.githubusercontent.com/81248/109205083-bbbf1400-776b-11eb-9fdd-058d0279e9e0.png)

fullofcaffeine commented on packages/e2e-tests/specs/editor/various/rtl.test.js:
```diff
@@ -21,6 +21,38 @@ describe( 'RTL', () => {
+			snippetScriptEl.text = 'document.dir = "rtl";';
+			newComponentsScriptEl.src = componentsScriptSrc;
+
+			//scripts.forEach( ( script ) => script.remove() );
```
Opsie.

fullofcaffeine committed:
Remove commented line

fullofcaffeine commented on packages/e2e-tests/specs/editor/various/rtl.test.js:
```diff
@@ -21,6 +21,37 @@ describe( 'RTL', () => {
+			const head = document.head;
+
+			const componentsScriptSrc = scripts.find( ( s ) =>
+				s.src.match( 'components/index.js' )
```
I don't quite like the fact that this ends up coupling this test with the build output (to a portion of the path to the compiled entry-point for `@wordpress/components`), but maybe it's fine as it won't change often ¯\_(ツ)_/¯

The alternative would be to just get all scripts without filtering, unshift the `snippetScriptEl`, and then re-add all of them again later. I tried that approach and it works, but then we would be re-evaluating all scripts, which is quite inefficient because we only really need to reload `@wordpress/components`. 

As much as I don't quite like the current approach, I think it provides the best trade-offs. Both approaches would be kind of hacky, but I don't see any other way of reproducing the crash caused by `rtlcss` from an e2e test. See my other attempts [here](https://github.com/WordPress/gutenberg/pull/29326#issuecomment-786139476).

fullofcaffeine replied:
Another thing is that there might be a possibility of a false-positive if the error is caught by the global error handler after the test finishes, which would mean the `window._loadedWithoutErrors` would still be `true`. While testing locally, the test consistently failed when I tested against `v10.0.0` though, but I think it's a possibility :/. I didn't find a good way to wait for the error without it creating a bottleneck in the case the error doesn't happen.

fullofcaffeine committed:
Be explicit and pass a regexp instead of a string

fullofcaffeine committed:
Simplify by removing the assertion and the onerror listener in the page context

Setting a variable to indicate an error happened is not needed as the uncaught error on the page will case the test to fail.

Also, the additional message added by having the expect was not really adding much value.

ockham commented:
Thanks @fullofcaffeine, it's great to see an e2e that actually manages to catch that issue!

That said, I agree that the way we have to remove and re-add `script` tags here is problematic (and very tightly coupled to implementation details). Furthermore, a point can be made that for consistency's sake, we should use the same kind of "RTL setting" for all tests in `rtl.test.js` -- preferably the kind that is as close as possible to real-world usage of GB, and is thus able to capture the most issues.

At this point, I think it'd be worth trying out to actually change the RTL test to install and activate an actual RTL language (for all the checks it runs). Let's maybe explore that in a different PR. Would you like to give that a try, or should I do that? :smile:

ockham commented:
Oh, there's also this plugin that I just learned about: https://wordpress.org/plugins/rtl-tester/

If we fail to make the e2e test install an RTL language, we might try to make it install that plugin instead :thinking:

ockham commented:
> Oh, there's also this plugin that I just learned about: https://wordpress.org/plugins/rtl-tester/
> 
> If we fail to make the e2e test install an RTL language, we might try to make it install that plugin instead

We can even take this a bit further:

Turns out we already maintain a few helper plugins for e2e tests in Gutenberg, see https://github.com/WordPress/gutenberg/tree/6449773e1340d38c2eb4e0555677df54df531040/packages/e2e-tests/plugins

Those are mostly used by the tests in https://github.com/WordPress/gutenberg/tree/6449773e1340d38c2eb4e0555677df54df531040/packages/e2e-tests/specs/editor/plugins.

The relevant logic in the RTL tester plugin is fairly simple: https://github.com/yoavf/RTL-Tester/blob/0f53a0468a70a3af4dbb736703481e9921d7caea/rtl-tester.php#L87-L91

So let's maybe just copy that code into a newly added `packages/e2e-tests/plugins/rtl.php`, and use that for our e2e test :smile:

fullofcaffeine commented:
Follow up: https://github.com/WordPress/gutenberg/pull/29598.
